client/lib/posts/post-edit-store.js

```diff
@@ -471,7 +471,23 @@ PostEditStore = {
 	},
 
 	isDirty: function() {
-		return _post !== _savedPost || _rawContent !== _initialRawContent;
+		debug(
+			'isDirty post?%s rawContent?%s initial=%s content=%s',
+			_post !== _savedPost,
+			_rawContent !== _initialRawContent,
+			_initialRawContent,
+			_rawContent
+		);
+		if ( _post !== _savedPost ) {
+			return true;
+		}
+		if ( _rawContent !== _initialRawContent ) {
+			return (
+				! isContentEmpty( _rawContent ) ||
```

Comment:
I have a feeling there may be some related issues around this line:

https://github.com/Automattic/wp-calypso/blob/76a4b7a/client/post-editor/post-editor.jsx#L753

Specifically, I was observing that when switching from HTML to Visual editors, it was setting the raw content without `<p>` tags. I'm not sure this is the same root cause as what you were observing, but I have a feeling that the solution here won't cover all cases (e.g. when switching from HTML mode to Visual mode where multiple paragraphs exist, does it show as dirty?).